src/libertem/udf/pca.py

```diff
@@ -0,0 +1,217 @@
+import collections
+
+import numpy as np
+
+from libertem.udf import UDF
+from libertem.udf.stddev import merge
+
+
+pca = collections.namedtuple('PCA', ['sum_im', 'var', 'N',  'n_components', 'singular_vals'])
+
+
+def flip_svd(U, V):
```

Comment:
This function is adapted from `sklearn`, right? Should be properly credited if this is the case (it is completely fine, as long as we comply with [their license](https://github.com/scikit-learn/scikit-learn/blob/master/COPYING)). Also, first part of the docstring in sklearn is kind of important: `"Sign correction to ensure deterministic output from SVD"` - it helps to understand the "why" of the function.

Reply:
Indeed yes. For citation, there is no "official" way of citing a code directly, other than in the form of a bibliography. Also, I couldn't find doi for scikit learn anywhere... Would providing a link to the specific code file (on github) be suffice?

Reply:
> For citation, there is no "official" way of citing a code directly
> [...] Would providing a link to the specific code file (on github) be suffice?

I _personally_ think a link should be fine (one could argue about the "threshold of originality" for copying 4 lines of code, but I'm not sure we want to go there :grin:) but the license says:

```
[...]
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  a. Redistributions of source code must retain the above copyright notice,
     this list of conditions and the following disclaimer.
[...]
```

So, following the license to the letter, we need to reproduce the license in full somewhere. A good example of this, actually in sklearn: https://github.com/scikit-learn/scikit-learn/blob/a50c03f975c4bbf306c5af1826f46ba5dc5230a0/sklearn/utils/_pprint.py - in that case, the whole file is from an external project. Maybe it makes sense to keep code you borrow/adapt from sklearn into a separate file, which has [the whole license](https://github.com/scikit-learn/scikit-learn/blob/master/COPYING) as a comment at the top. What do you think?

Reply:
Yeah that shouldn't be a problem at all. I've been writing code for NNMF on the side, and I was looking into sklearn code for reference as well (sklearn contained some cython and extra library stuff for nnmf so I didn't copy any yet except for maybe the general idea). If in the future such an issue of copying arises again in the future, having a separate file would ease the burden of citation